### ISSUE TITLE
Multi tenancy/connections via module caching problem.

### ISSUE THREAD

**Issue by ztripez** [wprl/baucis#298]
This will be a long post.

We have a mongo server with a database for each client, we created a rest api a couple of years ago using baucis version 0.6.25.

Now we have started a big refactor and rewrite of the system to upgrade it to express 4 and latest baucis. We also breaking a part old monolithic structure, database schemas/models etc, to be able to use it in workers and other places outside the rest server.

This have created some problem for us.

1. Baucis and mongoose must be initialized in the same module, (see my last issue: https://github.com/wprl/baucis/issues/297)
We fixed this with adding baucis as dependency to the database module and export it, express uses this instance of baucis. this works (in a way).
2. Multi tenancy dosn't works. Baucis ignores when we tell it to use a new connection and uses the first one only. We have done numerous testing and debugging. The database module have no problem with database switches but baucis ignores it.

(I've removed unrelated code and our hostnames/password from the code)

Module database-handler

    var mongoose = require('mongoose');
    var conn = mongoose.createConnection(config.get("database:uri")+"test");
    
    /*
    In this itteration of testing we use a static object, the same issue occurs with prototyped objects.
    We also have tried with a connection pool, with the same result.
    */
    
    function DatabaseHandler (identifier){
    
    var c = conn.useDb(identifier);
    	var returnObject = {
    		identifer:identifier,
    		connection:c,
                models: {
    					File: require('./model/File')(c),
    					User: require('./model/User')(c),
    					Scope:require('./model/Scopes')(c),
    				}
    	}
    	return returnObject;
    }
    
    DatabaseHandler.baucis = require('baucis');
    
    module.exports = DatabaseHandler;

Example of model module

    var UserSchema = require('./schema/UserSchema');
    module.exports = function (conn) {
      //to avoid mocca errors we only add the schema to the model if it isn't initialized previusly
      var model;
      try {
        model = conn.model("user")
      } catch (e) {
        model = conn.model("user", UserSchema)
      }
      return model;
    };

Our User schema

    var mongoose = require('mongoose');
    var Schema = mongoose.Schema;
    var bcrypt = require('bcrypt');
    
    /**
     * This paramter controlls how many iterations bcrypt will run on the password.
     * Changing this will make all currently saved passwords invalid. 
      */
    var SALT_WORK_FACTOR = XXXX;
    
    var UserSchema = mongoose.Schema({
        username : {
          type : String,
          required : true,
          unique : true,
          lowercase: true
        },
        email : {
          type : String,
          required : true,
          unique : true,
          lowercase: true
        },
        password : {
          type : String,
          required : true
        },
        name : {
          type : String,
          required : true
        },
        phone: {
    	    type : String
        },
        title: {
    	    type : String
        },
        department: {
    	    type : String
        },
        country: {
    	    type : String
        },
    	token : {
    		type: String,
    		require: false
    	},
    	tokendate : {
    		type: Date, 
    		default: Date.now,
    		require: false
    	} 
    });
    /*
     * hide fields when doing json, do not use select option
     */
    UserSchema.set('toJSON', {
        getters : true,
        virtuals : true,
        transform : function(doc, ret, options) {
          delete ret.password;
          delete ret.__v;
          delete ret.token;
          delete ret.tokendate;
          delete ret._id;
        }
    });
    
    // Bcrypt middleware
    UserSchema.pre('save', function(next) {
      var user = this;
      
      //Password hasn't changed
      if (!user.isModified('password')){
        return next();  
      }
      
      
      bcrypt.genSalt(SALT_WORK_FACTOR, function(err, salt) {
        if (err) return next(err);
        
        bcrypt.hash(user.password, salt, function(err, hash) {
          if (err){
	          return next(err);
	       }
          user.password = hash;
          next();
        });
      });
    });
    
    
    // Password verification
    UserSchema.methods.comparePassword = function(candidatePassword, cb) {
        bcrypt.compare(candidatePassword, this.password, function(err, isMatch) {
    	if (err)
    	    return cb(err);
    	cb(null, isMatch);
        });
    };
    module.exports = UserSchema;

Main rest server application:

server.js

    var db = require('database-handler');
    var express = require("express");
    var http = require('http');
	    /*
       		The boot sequence is complete and we can start the webserver
    	*/
    	
    	var app = express();
    	logger.info("Boot sequence complete, starting webserver");
    	app.enable('trust proxy');
    	var api = require("./routes/api");
    	app.use("/api/:identifier/",api);
    	app.set("port", config.get("PORT"));
    	
    	//Fetch database connection
    	app.param('identifier', function(req, res, next, id) {
    		req.db = db.database(id);
    		next();
    	});
    	
    	logger.info("Server listen to port %s",config.get("PORT"));
    	var server = http.createServer(app);
    	server.listen(config.get("PORT"));

API route:

    /**
     This route loads all controllers for clients
    */
        var express = require('express');
        var router = express.Router();
        var db = require('database-handler');
        
        router.use(function (req, res, next) {
        	var userController = require("../controller/UserController")(req.db);
        	//We have to force the location of the baucis routs.
        	req.app.use("/api/:identifier/",db.baucis());
        	next();
        })
        
        module.exports = router; 

The user controller

    var db = require('database-handler');
    module.exports = function (conn){
    	var controller = db.baucis.rest(conn.models.User);
		controller.query('get', function(req, res, next) {
			//Dump the baucis query connection objects  for debugging
			console.log(req.baucis.query.mongooseCollection.conn);
			return next();
		});
    	return controller;	
    }

* A call to api/client1/users will use the client1 database.
* A call to /api/client2/users will also use the client1 database.

However the req.db object is correct all the way.

We have also tried to get the connection directly from the module (not via req) but gives the same issue.

**Comment by ztripez** [comment]
Continued do some digging in baucis.

In Controller/configure.js line 69:
this model has the correct connection.

In Controller/query/ line 7
controller.model() returns a model with the old (wrong) connection.

**Comment by ztripez** [comment]
I did some digging in baucis and  some logged the database name of controller.model() in Controller/query/options.js
First call (client1 gave)

Line 29 : client1
Line 43 : client1
Line 60 : client1
Line 67 : client1
Line 74 : client1
Line 81 : client1
Line 97 : client1
Line 129 : client1
Line 140 : client1
Line 151 : client1
Line 174 : client1
Line 183 : client1

Second call (client2) gave:
Line 29 : client2
Line 29 : client2
Line 43 : client1
Line 60 : client1
Line 67 : client1
Line 74 : client1
Line 81 : client1
Line 97 : client1
Line 129 : client1
Line 140 : client1
Line 151 : client1
Line 174 : client1
Line 183 : client1

**Comment by wprl** [comment]
Thanks for digging deeper into this.  I will take a look.

**Comment by wprl** [comment]
I wonder if this may be caused by different parts of the code using different instances of the mongoose module… would you mind posting the output from `npm list`?

**Comment by ztripez** [comment]
Yes that's what i thought as well, so I did a major dependency overhaul.

this is my currenct npm list

```
backend-rest@1.0.0 /home/ztripez/code/backend-rest
├─┬ aws-sdk@2.2.11
│ ├── sax@0.5.3
│ ├── xml2js@0.2.8
│ └── xmlbuilder@0.4.2
├─┬ baucis@1.6.3
│ ├─┬ baucis-json@1.0.0
│ │ └─┬ event-stream@3.2.2
│ │   ├── duplexer@0.1.1
│ │   ├── from@0.1.3
│ │   ├── map-stream@0.1.0
│ │   ├── pause-stream@0.0.11
│ │   ├── split@0.3.3
│ │   ├── stream-combiner@0.0.4
│ │   └── through@2.3.8
│ ├── baucis-links@1.0.1
│ ├─┬ deco@0.12.3
│ │ └── requireindex@1.1.0
│ ├─┬ event-stream@3.3.2
│ │ ├── duplexer@0.1.1
│ │ ├── from@0.1.3
│ │ ├── map-stream@0.1.0
│ │ ├── pause-stream@0.0.11
│ │ ├── split@0.3.3
│ │ ├── stream-combiner@0.0.4
│ │ └── through@2.3.8
│ ├── rest-error@1.3.0
│ └── semver@4.3.6
├─┬ config-module-node@0.0.2
│ └─┬ nconf@0.8.2
│   ├── async@1.4.2
│   ├── ini@1.3.4
│   └─┬ yargs@3.29.0
│     ├── camelcase@1.2.1
│     ├─┬ cliui@3.0.3
│     │ ├─┬ string-width@1.0.1
│     │ │ ├─┬ code-point-at@1.0.0
│     │ │ │ └── number-is-nan@1.0.0
│     │ │ └─┬ is-fullwidth-code-point@1.0.0
│     │ │   └── number-is-nan@1.0.0
│     │ ├─┬ strip-ansi@3.0.0
│     │ │ └── ansi-regex@2.0.0
│     │ └── wrap-ansi@1.0.0
│     ├── decamelize@1.0.0
│     ├─┬ os-locale@1.4.0
│     │ └─┬ lcid@1.0.0
│     │   └── invert-kv@1.0.0
│     ├── window-size@0.1.2
│     └── y18n@3.2.0
├── connect-ensure-login@0.1.1
├─┬ cors@2.7.1
│ └── vary@1.1.0
├─┬ database-module-node@1.0.4
│ └─┬ bcrypt@0.8.5
│   ├── bindings@1.2.1
│   └── nan@2.0.5
├─┬ event-module-node@1.0.5
│ ├── logger-module-node@1.0.1
│ └─┬ winston@1.0.2
│   ├── async@1.0.0
│   ├── colors@1.0.3
│   ├── cycle@1.0.3
│   ├── eyes@0.1.8
│   ├── isstream@0.1.2
│   ├── pkginfo@0.3.1
│   └── stack-trace@0.0.9
├─┬ express@4.13.3
│ ├─┬ accepts@1.2.13
│ │ ├─┬ mime-types@2.1.7
│ │ │ └── mime-db@1.19.0
│ │ └── negotiator@0.5.3
│ ├── array-flatten@1.1.1
│ ├── content-disposition@0.5.0
│ ├── content-type@1.0.1
│ ├── cookie@0.1.3
│ ├── cookie-signature@1.0.6
│ ├─┬ debug@2.2.0
│ │ └── ms@0.7.1
│ ├── depd@1.0.1
│ ├── escape-html@1.0.2
│ ├── etag@1.7.0
│ ├─┬ finalhandler@0.4.0
│ │ └── unpipe@1.0.0
│ ├── fresh@0.3.0
│ ├── merge-descriptors@1.0.0
│ ├── methods@1.1.1
│ ├─┬ on-finished@2.3.0
│ │ └── ee-first@1.1.1
│ ├── parseurl@1.3.0
│ ├── path-to-regexp@0.1.7
│ ├─┬ proxy-addr@1.0.8
│ │ ├── forwarded@0.1.0
│ │ └── ipaddr.js@1.0.1
│ ├── qs@4.0.0
│ ├── range-parser@1.0.2
│ ├─┬ send@0.13.0
│ │ ├── destroy@1.0.3
│ │ ├─┬ http-errors@1.3.1
│ │ │ └── inherits@2.0.1
│ │ ├── mime@1.3.4
│ │ ├── ms@0.7.1
│ │ └── statuses@1.2.1
│ ├── serve-static@1.10.0
│ ├─┬ type-is@1.6.9
│ │ ├── media-typer@0.3.0
│ │ └─┬ mime-types@2.1.7
│ │   └── mime-db@1.19.0
│ ├── utils-merge@1.0.0
│ └── vary@1.0.1
├── logger-module-node@1.0.1
├─┬ mailer-module-node@0.0.2
│ └── node-uuid@1.4.3
├─┬ mongoose@4.2.2
│ ├── async@0.9.0
│ ├── bson@0.4.19
│ ├── hooks-fixed@1.1.0
│ ├── kareem@1.0.1
│ ├─┬ mongodb@2.0.46
│ │ ├── es6-promise@2.1.1
│ │ ├─┬ mongodb-core@1.2.19
│ │ │ └─┬ kerberos@0.0.15
│ │ │   └── nan@2.0.9
│ │ └─┬ readable-stream@1.0.31
│ │   ├── core-util-is@1.0.1
│ │   ├── inherits@2.0.1
│ │   ├── isarray@0.0.1
│ │   └── string_decoder@0.10.31
│ ├── mpath@0.1.1
│ ├── mpromise@0.5.4
│ ├─┬ mquery@1.6.3
│ │ ├── bluebird@2.9.26
│ │ └─┬ debug@2.2.0
│ │   └── ms@0.7.1
│ ├── ms@0.1.0
│ ├── muri@1.0.0
│ ├── regexp-clone@0.0.1
│ └── sliced@0.0.5
├─┬ mongoose-materialized@0.1.8
│ ├── async@0.9.2
│ └─┬ mongoose@3.6.20
│   ├── hooks@0.2.1
│   ├─┬ mongodb@1.3.19
│   │ ├── bson@0.2.2
│   │ └── kerberos@0.0.3
│   ├── mpath@0.1.1
│   ├─┬ mpromise@0.2.1
│   │ └── sliced@0.0.4
│   ├── ms@0.1.0
│   ├── muri@0.3.1
│   ├── regexp-clone@0.0.1
│   └── sliced@0.0.5
├─┬ mongoose-schema-extend@0.1.7
│ └── owl-deepcopy@0.0.4
├── mongoose-timestamp@0.3.0
├─┬ passport@0.3.0
│ ├── passport-strategy@1.0.0
│ └── pause@0.0.1
├─┬ passport-http@0.3.0
│ └── passport-strategy@1.0.0
├── q@1.4.1
├── queue-module-node@0.0.2
├── redis@2.1.0
├─┬ s3-streams@0.3.0
│ ├── bluebird@2.10.2
│ ├── lodash@3.10.1
│ └─┬ readable-stream@2.0.3
│   ├── core-util-is@1.0.1
│   ├── inherits@2.0.1
│   ├── isarray@0.0.1
│   ├── process-nextick-args@1.0.3
│   ├── string_decoder@0.10.31
│   └── util-deprecate@1.0.2
├── underscore@1.8.3
└─┬ winston@1.0.2
  ├── async@1.0.0
  ├── colors@1.0.3
  ├── cycle@1.0.3
  ├── eyes@0.1.8
  ├── isstream@0.1.2
  ├── pkginfo@0.3.1
  └── stack-trace@0.0.9
```

However, after I got a new error instead.

```
{
status: 400
name: "Bad Request"
message: "The requested document ID "scopes" is not a valid document ID (400)."
}
```
This is my scopes controller (worked (without multi tenancy) before):

```
var db = require('database-module-node');
var event = require('event-module-node');
var config = require('config-module-node');
var checkdelete = require('./middleware/checkdelete');
module.exports = function (conn){
	var controller = db.baucis.rest(conn.models.Scope);
	checkdelete(controller);
	return controller;	
}
```
conn.models.Scope is correct and everything.
I also tried to use bauics directly (not from the db module) with the same result.

**Comment by wprl** [comment]
I notice `mongoose-materialized` is using its own copy of mongoose — not sure but could be causing a problem… that is very strange that `req.params.id` would be altered.  I'm hard pressed to see how baucis is altering `req.params.id`. I believe it takes the value from express unaltered.

**Comment by ztripez** [comment]
I realized I forgot to answer. 
I got it to work and hopefully npm3 will solve this issue with its flatten dependency.

I also used an express parameter when mounting my baucis controllers

(the router is mapped to /api/:client)

First i used this inside the router
 req.app.use("/api/:client",db.baucis());

When i changed it to
req.app.use("/api/"+req.params.client,db.baucis());

Things started to work as intended.

Sorry for the late reply.
